python/tk_multi_launchapp/launch_apps.py

```diff
@@ -0,0 +1,651 @@
+# Copyright (c) 2016 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import os
+import re
+import sys
+
+import sgtk
+from sgtk import TankError
+
+from .prepare_apps import prepare_launch_for_engine
+
+shotgun_data = sgtk.platform.import_framework("tk-framework-shotgunutils", "shotgun_data")
+
+
+class BaseLauncher(object):
+    # documentation explaining how to reconfigure app paths
+    HELP_DOC_URL = "https://support.shotgunsoftware.com/entries/95443887#Setting%20up%20Application%20Paths"
+
+    def __init__(self):
+        # Retrieve the TK Application from the current bundle
+        self._tk_app = sgtk.platform.current_bundle()
+
+        # Store the current platform value
+        self._platform_name = {
+            "linux2": "linux", "darwin": "mac", "win32": "windows"
+        }[sys.platform]
+
+        # Initialize DCC information used during the launch process
+        self._set_app_info(None, None, None, None, None)
+
+    def _set_app_info(self, menu_name, app_engine, app_path, app_args, app_version):
```

Comment:
this is the part that i am not sure about. Maybe we can discuss. I think it makes sense that the `SingleConfigLauncher` has got member variables to track this - because it represents a single launch, but at this level i don't think there should be any internal state!

Instead, i think we should pass in all necessary state to both `_init_launch_command` (which passes it on to the callback `launch_app_from_entity`) and to `_launch_app`.

I think perhaps we could refactor these methods out into a `LaunchCallback` class or similar - that represents a single configuration and can contain member variables to describe the state etc. Do you think that could make sense? Let's discuss and apologies if i am missing some of the details! :)